### PR TITLE
XMR CLSAG support

### PR DESCRIPTION
**Pull request by ph4r05**
Adds support for CLSAG signature scheme that will be mandatory in the next Monero hard fork (13). 

Builds on top of the #942, I will rebase on the master once #942 is merged. Inspect pls only the last commit, others are part of #942.

- Thanks to @SarangNoether for notification on this matter so we can have CLSAG support prepared before rollout. We discussed it shortly and the transition should be very smooth.
- CLSAG is based on the paper https://eprint.iacr.org/2019/654.pdf (page 11-12 for construction)
- Reference implementation: https://github.com/SarangNoether/monero/commits/clsag-device
- CLSAG is quite similar to the MLSAG w.r.t. construction. 
- PR contains CLSAG unit test with CLSAG verifier implemented in tests. This tests positive and negative examples of CLSAG generated by the Trezor code. Plus the verifier is tested with the CLSAG generated in the reference C++ implementation from Monero.
- probably @onvej-sl will be interested in reviewing this :)

Core idea from the paper:
Screenshot 2020-04-12 at 18 08 40
Screenshot 2020-04-12 at 17 49 28

### CODE REVIEW

**Comment by prusnak** [comment]
When is the next Monero hardfork planned?

**Comment by ph4r05** [comment]
> When is the next Monero hardfork planned?

TBH I have no clue. Maybe @sarangNoether or @moneromooo-monero would know better? :)

**Comment by SarangNoether** [comment]
As far as I know, the block height for the next network upgrade has not yet been planned. I think it's likely that it will depend on when code, review, and auditing are completed for CLSAG. But it's not up to me.

**Comment by prusnak** [comment]
@SarangNoether Any rough estimate? Weeks? Months? Years?

**Comment by ph4r05** [comment]
The last time we discussed it months were mentioned.

File: core/src/apps/monero/xmr/mlsag.py [hunk 1]
```diff
@@ -273,6 +277,158 @@ def generate_mlsag(message, pk, xx, kLRki, index, dsRows, mg_buff):
...
+    :param a: mask from the pseudo output commitment; better name: pseudo_out_alpha
+    :param cout: pseudo output commitment; point, decoded; better name: pseudo_out_c
+    :param kLRki: used only in multisig, currently not implemented
+    :param index: specifies corresponding public key to the `in_sk` in the pubs array
+    :param mg_buff: buffer to store the signature to
+    """
+    cols = len(pubs)
+    if cols == 0:
+        raise ValueError("Empty pubs")
+
```
Comment (onvej-sl) [inline core/src/apps/monero/xmr/mlsag.py:299]: You should check `cols == len(cout)`.

File: core/src/apps/monero/signing/step_09_sign_input.py [hunk 2]
```diff
@@ -145,25 +145,23 @@ async def sign_input(
     utils.unimport_end(mods)
     state.mem_trace(3, True)
 
-    from apps.monero.xmr.serialize_messages.ct_keys import CtKey
-
     # Basic setup, sanity check
     index = src_entr.real_output
-    input_secret_key = CtKey(dest=spend_key, mask=crypto.decodeint(src_entr.mask))
+    input_secret_key = (spend_key, crypto.decodeint(src_entr.mask))
```
[the remark this answers is not in the document]
Comment (ph4r05) [reply core/src/apps/monero/signing/step_09_sign_input.py:150]: Aah I thought we can remove this as it is used only at one place :D 
Do you want it back?

File: core/src/apps/monero/xmr/mlsag.py [hunk 3]
```diff
@@ -273,6 +277,158 @@ def generate_mlsag(message, pk, xx, kLRki, index, dsRows, mg_buff):
...
+    tmp_sc = crypto.new_scalar()
+    tmp = crypto.new_point()
+    tmp_bf = bytearray(32)
+
+    crypto.hash_to_point_into(H, P[index])
+    crypto.scalarmult_into(sI, H, p)  # I = p*H
+    crypto.scalarmult_into(D, H, z)  # D = z*H
+    crypto.sc_mul_into(tmp_sc, z, crypto.sc_inv_eight())  # -8*z
+    crypto.scalarmult_into(sD, H, tmp_sc)  # sig.D = -8*z*H
```
[the remark this answers is not in the document]
Comment (ph4r05) [reply core/src/apps/monero/xmr/mlsag.py:333]: Ah, sure!

**Comment by ph4r05** [comment]
Btw - thanks for super fast review!

File: core/src/apps/monero/xmr/mlsag.py (hunk 1 above)
Comment (ph4r05) [reply core/src/apps/monero/xmr/mlsag.py:299]: I am adding the typing info, cout is Ge25519

File: core/src/apps/monero/signing/step_09_sign_input.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (ph4r05) [reply core/src/apps/monero/signing/step_09_sign_input.py:150]: its back

**Comment by ph4r05** [comment]
Closing as CLSAG is now part of the https://github.com/trezor/trezor-firmware/pull/942

**Pull request closed without merging**